compiler/rustc_ast_passes/src/errors.rs

```diff
@@ -823,3 +823,10 @@ pub(crate) struct DuplicatePreciseCapturing {
     #[label]
     pub bound2: Span,
 }
+
+#[derive(Diagnostic)]
+#[diag(ast_passes_extern_without_abi)]
+pub(crate) struct MissingAbi {
+    #[primary_span]
+    pub span: Span,
+}
```

Comment:
The hard error should have a suggestion too, in case someone changes the edition string by hand.

Reply:
I have mixed feelings about a suggestion in the new edition.  One purpose of making this a hard error is that we don't want to suggest that "C" is the default any longer (and we may want to later consider making "Rust" the default).  We do machine-applicable suggestions when we're really sure that's the right fix, but once we move away from this default, it's no longer actually clear that is what the user meant.

If we want to add some help text here telling the user that old editions inferred a default and that this is no longer true, and so the user needs to pick one explicitly, that'd of course be OK.  If we want to go further and add a machine-applicable fix to this hard error, then I'd want a lang decision here to that effect.

Reply:
Help text sounds good, it solves the problem of "user updates edition without thinking, doesn't understand the new error".

Reply:
I have added the help text. @Nadrieril @traviscross